fix: consider pod density in memory overhead calc

mKeRix opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
docs:            <-- Documentation change that does not impact code
chore:           <-- Metadata changes such as dependency update or configuration files
test:            <-- Test changes that do not impact behavior
perf:            <-- Code changes that improve performance but do not impact behavior
BREAKING CHANGE: <-- Include if your change includes a backwards incompatible change
-->

Fixes #1959

**Description**
This change will take the pod density setting in account when calculating memory overhead. This is important to avoid scheduling mistakes and endless loops like described in the linked issue.

**How was this change tested?**

* Real EKS cluster using the manifests from the issue
* Newly added automated tests

See new logs when testing against the reproducible setup (as expected, the c5n.xlarge instance type is now considered to be too small):

```console
2022-06-21T07:09:04.372Z	DEBUG	controller.provisioning	Relaxing soft constraints for pod since it previously failed to schedule, removing: spec.topologySpreadConstraints = {"maxSkew":1,"topologyKey":"topology.kubernetes.io/zone","whenUnsatisfiable":"ScheduleAnyway","labelSelector":{"matchLabels":{"app":"inflate"}}}	{"commit": "153d333"}
2022-06-21T07:09:04.376Z	DEBUG	controller.provisioning	Relaxing soft constraints for pod since it previously failed to schedule, adding: toleration for PreferNoSchedule taints	{"commit": "153d333"}
2022-06-21T07:09:04.379Z	ERROR	controller.provisioning	Could not schedule pod, incompatible with provisioner "default", no instance type satisfied resources {"cpu":"224m","memory":"8791517761","pods":"1"} and requirements karpenter.sh/capacity-type In [on-demand spot], kubernetes.io/arch In [amd64 arm64], node.kubernetes.io/instance-type In [c5n.xlarge], karpenter.sh/provisioner-name In [default]; incompatible with provisioner "long-running", did not tolerate idealo.de/long-running=true:NoSchedule	{"commit": "153d333", "pod": "inflate/inflate-5bb76d8549-rprbr"}
```

Once you remove the `nodeSelector` it picks a list of instance types correctly:

```console
2022-06-21T07:11:34.973Z	DEBUG	controller.provisioning.cloudprovider	Created launch template, Karpenter-terratest-u1q5gG-10432077704041859839	{"commit": "153d333", "provisioner": "default"}
2022-06-21T07:11:36.890Z	INFO	controller.provisioning.cloudprovider	Launched instance: i-0e752dc6a294469a1, hostname: ip-172-16-66-212.eu-central-1.compute.internal, type: c6i.2xlarge, zone: eu-central-1b, capacityType: spot	{"commit": "153d333", "provisioner": "default"}
2022-06-21T07:11:36.922Z	INFO	controller.provisioning	Created node with 1 pods requesting {"cpu":"559m","memory":"9037933121","pods":"8"} from types m6i.xlarge, m5zn.xlarge, t3a.xlarge, t3.xlarge, m6a.xlarge and 193 other(s)	{"commit": "153d333", "provisioner": "default"}
```

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # <!-- issue number -->
- [x] No

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Fixed calculation of allocatable memory when not using ENI-based pod density
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dewjam commented:
Hey @mKeRix ,
Thanks for submitting the PR.  I'm going to pull down your code and do some testing of my own.  In the meantime, can you work on adding some automated tests?

dewjam reviewed (changes requested):
Per my previous comment, please add some automated testing.

mKeRix commented on pkg/cloudprovider/aws/instancetype.go:
```diff
@@ -270,7 +269,7 @@ func (i *InstanceType) computeOverhead(vmMemOverhead float64) v1.ResourceList {
 			(int64(math.Ceil(float64(memory.Value())*vmMemOverhead/1024/1024)))+
 				// kube-reserved
-				((11*i.eniLimitedPods())+255)+
+				((11*pods.Value())+255)+
```
Interesting! Bottlerocket seems to behave differently then, as you can see [here](https://github.com/bottlerocket-os/bottlerocket/blob/5d69ebc1fc76a6032b677c9f3941bfa58221f3af/packages/kubernetes-1.22/kubelet-config#L71) and [here](https://github.com/bottlerocket-os/bottlerocket/blob/862c4fa3b9c09a8f6696c360a63ba644b3caaf35/sources/api/schnauzer/src/helpers.rs#L921). When max-pods is set in the user data it seems to always configure `max_pods*11+255 Mi` for the kube reserved memory. Without it, it will take the ENI based pod limits (or 110 as default if it doesn't have the data, e.g. for new instance types).

While I agree that this is a waste on small instance types, I would argue that Karpenter should calculate in the same ways as the AMIs do. Maybe a future update could consider a more dynamic max pods setting for non ENI based pod density (e.g. scale based on available vCPUs)? The code from this PR should already work out of the box then.

I do wonder what an appropriate way forward would be. Since these values are pre-calculated I don't think I can build a switch statement based on the AMI family here. I can add some code to the AL2 AMI user data generation that will override the reserved memory with the more appropriate one maybe? That would have the additional benefit of working around the issue you mentioned. What do you think?

suket22 replied:
Ah I see - so the reason you're seeing a difference in behavior is because you're using Bottlerocket and kubeReserved is being calculated differently by that AMI (compared to AL2). I see that your difference in allocatable as per #1959 was `84.60 Mebibytes` (`9037933121 - 8949215232` bytes). But if I'm calculating correctly, I'd have expected a larger diff since 893Mb is what we're calculating (11 * 58 + 255) in the binpacking logic,  and the BottleRocket AMI is doing 1465Mb.

I agree that we should match the calculation of the AMI in the bin-packing logic and I'm open to changing the way we bin-pack for Bottlerocket nodes. 

In th elong term, I think we want the maxPods memory calculation to be a function of the memory on the instanceType and use something [like this](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#memory_cpu).

mKeRix replied:
I just had another look with a fresh set of eyes and realized that my thoughts from the last comment were way off. I'll implement the calculation adjustment based on AMI family as you suggested and will report back. Thanks for all your input!

mKeRix commented:
@dewjam fyi: I'm happy to look at adding some test code - before investing more time I'd like to wait for some input on the direction of this PR (see other comment) though.

mKeRix commented:
Hey @dewjam & @suket22, I've updated the PR with an implementation that takes the AMIFamily into account when calculating the pod memory overhead and added some tests. I'd appreciate a new review. :)

mKeRix commented:
I just rebased the branch on the latest changes from main and simplified the tests. Is there anything else I can do to help this get merged?

ellistarn commented:
Some folks are out for the long weekend @mKeRix . Thanks for your patience and we'll get to this as soon as we can.

mKeRix commented:
Hey there, just wanted to check in again to see if we can get this moving? We'd love to see this change in an upstream release.

ellistarn commented:
@suket22 @dewjam

suket22 commented on pkg/cloudprovider/aws/amifamily/resolver.go:
```diff
@@ -75,6 +75,7 @@ type AMIFamily interface {
 	DefaultMetadataOptions() *v1alpha1.MetadataOptions
 	EphemeralBlockDevice() *string
 	EphemeralBlockDeviceOverhead() resource.Quantity
+	ENILimitedMemoryOverhead() bool
```
I think `MaxPods(i *InstanceType) int` might've been more flexible so if there's further divergence in how different AMIs are calculating MaxPods, we can just modify the AMI specific calculation and leave the others. 

Not blocking on this PR though, we can re-work that later 👍

mKeRix replied:
For this specific issue something like `MemoryOverhead(i *InstanceType) resource.Quantity` might be better (as the difference is in the calculation of the overhead, the max pods match correctly in the current implementation). Nonetheless, I think both functions would be a great addition and would give more flexibility in the implementation. 👍

suket22 commented on pkg/cloudprovider/aws/suite_test.go:
```diff
@@ -370,6 +370,50 @@ var _ = Describe("Allocation", func() {
+					Expect(err).To(BeNil())
+					it := NewInstanceType(injection.WithOptions(ctx, opts), instanceInfo["m5.xlarge"], 0, provider, nil)
+					overhead := it.Overhead()
+					Expect(overhead.Memory().String()).To(Equal("3073Mi"))
```
I was very confused reading `3073Mi` tbh, because that couldn't be right for an m5.xlarge. 
Did a little debugging, and the problem is in our [mocked EC2 response here](https://github.com/aws/karpenter/blob/main/pkg/cloudprovider/aws/fake/ec2api.go#L375). That should be `15` and not `60` which gives you an actual memory overhead of `1093Mi` instead for an m5.xlarge. 

Would you mind switching that 60 to a 15 and updating this PR?

mKeRix replied:
Good catch! I've pushed a commit that updates the mock response and corresponding assertions. I did manually calculate the values to verify during the implementation, but honestly never thought about the logic of the two values. This should make the intention much more clear.

suket22 commented on pkg/cloudprovider/aws/suite_test.go:
```diff
@@ -370,6 +370,50 @@ var _ = Describe("Allocation", func() {
+					Expect(err).To(BeNil())
+					it := NewInstanceType(injection.WithOptions(ctx, opts), instanceInfo["m5.xlarge"], 0, provider, nil)
+					overhead := it.Overhead()
+					Expect(overhead.Memory().String()).To(Equal("1665Mi"))
```
First I was confused how we aren't limited by ENIs yet our overhead for BR dropped from `3073Mi` to `1665Mi`. 
I think it makes sense now since it should rather be an increase from `1093Mi` to `1665Mi`.

suket22 reviewed (commented):
Apologies for the wait, and thank you for making this change.
Mostly LGTM, just need to flip a couple numbers in the tests :)

mKeRix commented:
@suket22 I've updated the PR according to your comments - would appreciate another review. :)

dewjam commented:
Hey @mKeRix ,
I'm testing your PR now, just to confirm some behaviors.  I should be able to give you an update today.  Thanks for your patience!
